DOCS-2703: Add support for camera with esp32 (#3294)
viamrobotics/docs: docs/components/camera/esp32-camera.md

---

> title: "Configure an esp32-camera (Micro-RDK)"

Author: ```suggestion
title: "Configure an esp32-camera (`viam-micro-server`)"
```

> title: "Configure an esp32-camera (`viam-micro-server`)"

Reviewer: ```suggestion
title: "Configure an esp32-camera (viam-micro-server)"
```

linkTitle: "esp32-camera"
weight: 33
type: "docs"
description: "Configure a camera connected to an esp32 board, initialized and configured using esp-idf."
images: ["/icons/components/camera.svg"]
tags: ["camera", "components", "micro-RDK"]
component_description: "An `OV2640` or `OV3660` camera connected to an esp32 board."
micrordk_component: true
toc_hide: true
# SMEs: Matt Perez, micro-RDK team
---

> `esp32-camera` is the camera model for cameras connected to an `esp32` board, which are initialized and configured using the [`viam-micro-server` Development Setup](/installation/micro-rdk-dev/) (ESP-IDF).
> If the camera drivers are one of the following two, the camera will work with the `esp32-camera` model:
> 
> - `OV2640`
> 
>   - [Datasheet](https://www.uctronics.com/download/OV2640_DS.pdf)
>   - 1600 x 1200, Color, ¼” lens
>   - You can use a cam ribbon adapter to connect to your `esp32` board
> 
> - `OV3660`: an [m5 camera timer module](https://docs.m5stack.com/en/unit/timercam_f)
>   - [Datasheet](https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/unit/OV3660_CSP3_DS_1.3_sida.pdf)
>   - 2048 x 1536, Color, ⅕” lens
> 

Reviewer: they are not configured using the dev setup... they are configured on this page. the micro-server provides the drivers for this. There is no operating system on the esp32 
```suggestion
`esp32-camera` is the camera model the following two cameras:

- `OV2640`

  - [Datasheet](https://www.uctronics.com/download/OV2640_DS.pdf)
  - 1600 x 1200, Color, ¼” lens
  - You can use a cam ribbon adapter to connect to your `esp32` board

- `OV3660`: an [m5 camera timer module](https://docs.m5stack.com/en/unit/timercam_f)
  - [Datasheet](https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/unit/OV3660_CSP3_DS_1.3_sida.pdf)
  - 2048 x 1536, Color, ⅕” lens

```

Author: @mattjperez this is meant to be used with more than just these two cameras right? What do you want to say/do you want to keep it open?

Reviewer 2: correct. This driver is basically a wrapper around espressif's [esp32-camera drivers](https://github.com/espressif/esp32-camera), with a subset of the functionality. eventually we'd like to expose almost all of the functionality in a viam way.

Reviewer: ```suggestion
`esp32-camera` is the camera model that supports all cameras that work with espressif's [esp32-camera drivers](https://github.com/espressif/esp32-camera) including:

- `OV2640`

  - [Datasheet](https://www.uctronics.com/download/OV2640_DS.pdf)
  - 1600 x 1200, Color, ¼” lens
  - You can use a cam ribbon adapter to connect to your `esp32` board

- `OV3660`: an [m5 camera timer module](https://docs.m5stack.com/en/unit/timercam_f)
  - [Datasheet](https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/unit/OV3660_CSP3_DS_1.3_sida.pdf)
  - 2048 x 1536, Color, ⅕” lens

```

Author: changed

For example, the `OV2640` with a ribbon cable connected to an ESP32:

{{< imgproc src="/components/camera/esp32-camera-2640.png" alt="Fake Camera on the ESP32" resize="300x" >}}

## Software requirements

> To use this model, you must follow the [`viam-micro-server` Development Setup](/installation/micro-rdk-dev/), which enables you to install and activate the ESP-IDF.

Reviewer: same comment as on the fake model

> At the step [Generate a new project from the `viam-micro-server` template](/installation/micro-rdk-dev/#generate-a-new-project-from-the-micro-RDK-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.

Author: ```suggestion
At the step [Generate a new project from the `viam-micro-server` template](/installation/micro-rdk-dev/#generate-a-new-project-from-the-micro-rdk-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.
```

Finish the [`viam-micro-server` Development Setup](/installation/micro-rdk-dev/) and return to this guide.

{{< alert title="Info" color="info" >}}

> The`esp32-camera` camera model is not currently available as a built-in option in [the Viam app](https://app.viam.com), so you cannot use **Builder** mode to configure this board.

Reviewer: ```suggestion
The`esp32-camera` camera model is not currently available as a built-in option in [the Viam app](https://app.viam.com), so you cannot use **Builder** mode to configure this camera.
```

but also I think we can skip this comment. Given that there's no builder mode tab, that kind of conveys the same message

{{< /alert >}}

{{< tabs name="Configure a esp32-camera" >}}
{{% tab name="JSON Template" %}}

Navigate to the **CONFIGURE** tab of your machine's page in [the Viam app](https://app.viam.com).
Select **Builder** mode.
Copy and paste the following JSON into your existing machine configuration in the `"components"` level:

> ```json {class="line-numbers linkable-line-numbers"}

Reviewer: oh wow. Are the pins on the other tabs always those pins? If so maybe we can prefill them...

Author: @mattjperez

Reviewer 2: it's dependent on the esp32 you're using **and** the camera you're pairing it with, or more specifially, what pins the ribbon cable of the camera gets mapped to when plugged in.

    "pin_pwdn": -1,
    "pin_reset": -1,
    "pin_xclk": 21,
    "pin_sccb_sda": 26,
    "pin_sccb_scl": 27,
    "pin_d7": 35,
    "pin_d6": 34,
    "pin_d5": 39,
    "pin_d4": 36,
    "pin_d3": 19,
    "pin_d2": 18,
    "pin_d1": 5,
    "pin_d0": 4,
    "pin_vsync": 25,
    "pin_href": 23,
    "pin_pclk": 22,
    "xclk_freq_hz": 20000000,
    "ledc_timer": 1,
    "ledc_channel": 1
  }
}
```

{{% /tab %}}
{{% tab name="JSON Example: OV3660" %}}

```json {class="line-numbers linkable-line-numbers"}
{
  "name": "my-esp32camera",
  "namespace": "rdk",
  "type": "camera",
  "model": "esp32-camera",
  "attributes": {
    "pin_d4": 39,
    "jpeg_quality": 32,
    "frame_size": 5,
    "pin_d5": 18,
    "pin_d3": 5,
    "pin_d6": 36,
    "pin_vsync": 22,
    "ledc_timer": 1,
    "pin_d7": 19,
    "pin_sccb_sda": 25,
    "pin_href": 26,
    "pin_sccb_scl": 23,
    "sccb_i2c_port": -1,
    "pin_d1": 35,
    "pin_d0": 32,
    "pin_xclk": 27,
    "pin_reset": 15,
    "pin_pclk": 21,
    "pin_d2": 34,
    "xclk_freq_hz": 20000000,
    "ledc_channel": 1
  }
}
```

{{% /tab %}}
{{< /tabs >}}

The following attributes are available for `esp32-camera` cameras:

<!-- prettier-ignore -->

> | Name | Type | Required? | Description |

Reviewer: This doesn't make sense. Generally we only list a default if they are optional. Because it means those values will be used otherwise. Are these suggestions? If so we should make that clearer

Reviewer: I see you have a comment about this. That means they are not required actually. Let's put the note about the datasheet above the table I think

Author: changed to optional and moved note!

For example, the `OV2640` with a ribbon cable connected to an ESP32:

{{< imgproc src="/components/camera/esp32-camera-2640.png" alt="Fake Camera on the ESP32" resize="300x" >}}

{{< alert title="Software requirements" color="note" >}}
To use this model, you must follow the [Micro-RDK Development Setup](/installation/micro-rdk-dev/), which enables you to install and activate the ESP-IDF.
At the step [Generate a new project from the micro-RDK template](/installation/micro-rdk-dev/#generate-a-new-project-from-the-micro-rdk-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.
Finish the [Micro-RDK Development Setup](/installation/micro-rdk-dev/) and return to this guide.
{{< /alert >}}

{{< alert title="Info" color="info" >}}

The`esp32-camera` camera model is not currently available as a built-in option in [the Viam app](https://app.viam.com), so you cannot use **Builder** mode to configure this camera.

{{< /alert >}}

{{< tabs name="Configure a esp32-camera" >}}
{{% tab name="JSON Template" %}}

Navigate to the **CONFIGURE** tab of your machine's page in [the Viam app](https://app.viam.com).

> Select **Builder** mode.

Author: ```suggestion
Select **JSON** mode.
```

> Copy and paste the following JSON into your existing machine configuration in the `"components"` level:

Reviewer: This doesn't make sense

Author: ah yes it needs to be JSON mode, changed

  "type": "camera",
  "model": "esp32-camera",
  "attributes": {
    "pin_pwdn": -1,
    "pin_reset": -1,
    "pin_xclk": 21,
    "pin_sccb_sda": 26,
    "pin_sccb_scl": 27,
    "pin_d7": 35,
    "pin_d6": 34,
    "pin_d5": 39,
    "pin_d4": 36,
    "pin_d3": 19,
    "pin_d2": 18,
    "pin_d1": 5,
    "pin_d0": 4,
    "pin_vsync": 25,
    "pin_href": 23,
    "pin_pclk": 22,
    "xclk_freq_hz": 20000000,
    "ledc_timer": 1,
    "ledc_channel": 1
  }
}
```

{{% /tab %}}
{{% tab name="JSON Example: OV3660" %}}

```json {class="line-numbers linkable-line-numbers"}
{
  "name": "my-esp32camera",
  "namespace": "rdk",
  "type": "camera",
  "model": "esp32-camera",
  "attributes": {
    "pin_d4": 39,
    "jpeg_quality": 32,
    "frame_size": 5,
    "pin_d5": 18,
    "pin_d3": 5,
    "pin_d6": 36,
    "pin_vsync": 22,
    "ledc_timer": 1,
    "pin_d7": 19,
    "pin_sccb_sda": 25,
    "pin_href": 26,
    "pin_sccb_scl": 23,
    "sccb_i2c_port": -1,
    "pin_d1": 35,
    "pin_d0": 32,
    "pin_xclk": 27,
    "pin_reset": 15,
    "pin_pclk": 21,
    "pin_d2": 34,
    "xclk_freq_hz": 20000000,
    "ledc_channel": 1
  }
}
```

{{% /tab %}}
{{< /tabs >}}

{{% alert title="Note" color="note" %}}

> While the following attributes marked as **Optional** do have defaults, it is still recommended that you configure them to your datasheet and device's needs (which likely do not align with the defaults) in order to get the camera stream.

Reviewer: @mattjperez can you confirm this? The example ones worked fine for my esp32-wrover-dev. Are they really likely to not align?

Reviewer 2: until we get more boards (with ribbon ports) and cameras and test, I don't know how likely it is to be honest. It might be standardized or it might not. Either way, users should double check so they don't accidentally break their board and/or camera.

Reviewer 2: maybe change from `which likely do not align with defaults` to `which may not align with the defaults, possibly damaging your board and/or camera`. the damage part is kind of a given in this space, but not sure how much that has been emphasized up to this point

Author: done

| `pin_d2` | int | Optional | GPIO pin for camera D2 line. <br> Default: 18 |
| `pin_d1` | int | Optional | GPIO pin for camera D1 line. <br> Default: 5 |
| `pin_d0` | int | Optional | GPIO pin for camera D0 line. <br> Default: 4 |
| `pin_vsync` | int | Optional | GPIO pin for camera VSYNC line. <br> Default: 25 |
| `pin_href` | int | Optional | GPIO pin for camera HREF line. <br> Default: 23 |
| `pin_pclk` | int | Optional | GPIO pin for camera PLCK line. <br> Default: 22 |
| `xclk_freq_hz` | int | Optional | Frequency of XCLK signal, in Hz. <br> Experimental: Set to 16MHz on ESP32-S2 or ESP32-S3 to enable EDMA mode. <br> Default: 20000000 |
| `ledc_timer` | int | Optional | LEDC timer to generate XCLK. <br> Default: 1 |
| `ledc_channel` | int | Optional | LEDC channel to generate XCLK. <br> Default: 1 |
| `frame_size` | int | Optional | Size of the output image. <br> Default: 1 |
| `jpeg_quality` | int | Optional | Quality of JPEG output. Lower means higher quality. <br> Range: 0-63 <br> Default: 32 |

## View the camera stream

{{< readfile "/static/include/components/camera-view-camera-stream.md" >}}

## Next steps

> Configure [data capture](/services/data/capture/) and [cloud sync](/services/data/cloud-sync/) for `viam-micro-server` to capture images from the camera stream and store them in the Viam cloud.

Reviewer: Offer it as an option - because it's not necessarily the next step. And better to link to the how to

```suggestion
If you need to capture image data and sync it to the cloud, see:

{{< cards >}}
{{% card link="/how-tos/image-data/" %}}
{{< /cards >}}
```

tags: ["camera", "components", "micro-RDK"]
component_description: "An `OV2640` or `OV3660` camera connected to an esp32 board."
micrordk_component: true
toc_hide: true
# SMEs: Matt Perez, micro-RDK team
---

`esp32-camera` is the camera model that supports all cameras that work with Espressif's [esp32-camera drivers](https://github.com/espressif/esp32-camera) including:

- `OV2640`

  - [Datasheet](https://www.uctronics.com/download/OV2640_DS.pdf)
  - 1600 x 1200, Color, ¼” lens
  - You can use a cam ribbon adapter to connect to your `esp32` board

- `OV3660`: an [m5 camera timer module](https://docs.m5stack.com/en/unit/timercam_f)
  - [Datasheet](https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/datasheet/unit/OV3660_CSP3_DS_1.3_sida.pdf)
  - 2048 x 1536, Color, ⅕” lens

For example, the `OV2640` with a ribbon cable connected to an ESP32:

{{< imgproc src="/components/camera/esp32-camera-2640.png" alt="Fake Camera on the ESP32" resize="300x" >}}

{{< alert title="Software requirements" color="note" >}}
To use this model, you must follow the [Micro-RDK Development Setup](/installation/micro-rdk-dev/), which enables you to install and activate the ESP-IDF.

> At the step [Generate a new project from the micro-RDK template](/installation/micro-rdk-dev/#generate-a-new-project-from-a-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.

Author: ```suggestion
At the step [Generate a new project from a template](/installation/micro-rdk-dev/#generate-a-new-project-from-a-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.
```

> At the step [Generate a new project from the micro-RDK template](/installation/micro-rdk-dev/#generate-a-new-project-from-the-micro-rdk-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.

Author: ```suggestion
At the step [Generate a new project from the micro-RDK template](/installation/micro-rdk-dev/#generate-a-new-project-from-a-template) where you create a new project with `cargo generate`, select the option to include camera module traits when prompted.
```

Finish the [Micro-RDK Development Setup](/installation/micro-rdk-dev/) and return to this guide.
{{< /alert >}}

{{< alert title="Info" color="info" >}}

The`esp32-camera` camera model is not currently available as a built-in option in [the Viam app](https://app.viam.com), so you cannot use **Builder** mode to configure this camera.

{{< /alert >}}

{{< tabs name="Configure a esp32-camera" >}}
{{% tab name="JSON Template" %}}

Navigate to the **CONFIGURE** tab of your machine's page in [the Viam app](https://app.viam.com).
Select **JSON** mode.

> Copy and paste the following JSON into your existing machine configuration in the `"components"` level:

Reviewer: ```suggestion
Copy and paste the following JSON into your existing machine configuration into the `"components"` array:
```

  "type": "camera",
  "model": "esp32-camera",
  "attributes": {
    "pin_pwdn": -1,
    "pin_reset": -1,
    "pin_xclk": 21,
    "pin_sccb_sda": 26,
    "pin_sccb_scl": 27,
    "pin_d7": 35,
    "pin_d6": 34,
    "pin_d5": 39,
    "pin_d4": 36,
    "pin_d3": 19,
    "pin_d2": 18,
    "pin_d1": 5,
    "pin_d0": 4,
    "pin_vsync": 25,
    "pin_href": 23,
    "pin_pclk": 22,
    "xclk_freq_hz": 20000000,
    "ledc_timer": 1,
    "ledc_channel": 1
  }
}
```

{{% /tab %}}
{{% tab name="JSON Example: OV3660" %}}

```json {class="line-numbers linkable-line-numbers"}
{
  "name": "my-esp32camera",
  "namespace": "rdk",
  "type": "camera",
  "model": "esp32-camera",
  "attributes": {
    "pin_d4": 39,
    "jpeg_quality": 32,
    "frame_size": 5,
    "pin_d5": 18,
    "pin_d3": 5,
    "pin_d6": 36,
    "pin_vsync": 22,
    "ledc_timer": 1,
    "pin_d7": 19,
    "pin_sccb_sda": 25,
    "pin_href": 26,
    "pin_sccb_scl": 23,
    "sccb_i2c_port": -1,
    "pin_d1": 35,
    "pin_d0": 32,
    "pin_xclk": 27,
    "pin_reset": 15,
    "pin_pclk": 21,
    "pin_d2": 34,
    "xclk_freq_hz": 20000000,
    "ledc_channel": 1
  }
}
```

{{% /tab %}}
{{< /tabs >}}

{{% alert title="Note" color="note" %}}

> While the following attributes marked as **Optional** do have defaults, it is still recommended that you configure them to your datasheet and device's needs (which likely do not align with the defaults) in order to get the camera stream.

Reviewer: ```suggestion
While the following attributes marked as **Optional** do have defaults, it is recommended that you configure them according to your datasheet as your device may not align with the defaults which could cause damage to your board or camera.
```
